pkg/operator/management/management_state_controller.go

```diff
@@ -0,0 +1,158 @@
+	clusterOperatorGetter configv1client.ClusterOperatorsGetter
+	eventRecorder         events.Recorder
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewOperatorManagementStateController(
+	name string,
+	clusterOperatorClient configv1client.ClusterOperatorsGetter,
+	operatorStatusProvider operatorv1helpers.OperatorClient,
+	recorder events.Recorder,
+) *ManagementStateController {
+	c := &ManagementStateController{
+		operatorName:          name,
+		clusterOperatorGetter: clusterOperatorClient,
+		operatorClient:        operatorStatusProvider,
+		eventRecorder:         recorder,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ManagementStateController-"+name),
+	}
+
+	operatorStatusProvider.Informer().AddEventHandler(c.eventHandler())
+	// TODO watch clusterOperator.status changes when it moves to openshift/api
+
+	return c
+}
+
+func (c ManagementStateController) sync() error {
+	detailedSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if apierrors.IsNotFound(err) {
+		c.eventRecorder.Warningf("StatusNotFound", "Unable to determine current operator status for %s", c.operatorName)
+		return nil
+	}
+
+	originalOperatorObj, err := c.clusterOperatorGetter.ClusterOperators().Get(c.operatorName, metav1.GetOptions{})
+	if err != nil {
+		// Wait for the status controller to create status for us
+		glog.Infof("Unable to get operator status: %v", err)
+		return err
+	}
+
+	operatorObj := originalOperatorObj.DeepCopy()
+	operatorObj.Status = configv1.ClusterOperatorStatus{}
+
+	if IsOperatorAlwaysManaged() && detailedSpec.ManagementState == operatorv1.Unmanaged {
+		configv1helpers.SetStatusCondition(&operatorObj.Status.Conditions, configv1.ClusterOperatorStatusCondition{
+			Type:   "ManagementStateFailing",
+			Status: configv1.ConditionTrue,
+			Reason: fmt.Sprintf("Unmanaged is not supported for %s operator", c.operatorName),
+		})
+	}
+
+	if IsOperatorNotRemovable() && detailedSpec.ManagementState == operatorv1.Removed {
+		configv1helpers.SetStatusCondition(&operatorObj.Status.Conditions, configv1.ClusterOperatorStatusCondition{
+			Type:   "ManagementStateFailing",
+			Status: configv1.ConditionTrue,
+			Reason: fmt.Sprintf("Removed is not supported for %s operator", c.operatorName),
+		})
+	}
+
+	if IsOperatorUnknownState(detailedSpec.ManagementState) {
+		configv1helpers.SetStatusCondition(&operatorObj.Status.Conditions, configv1.ClusterOperatorStatusCondition{
+			Type:   "ManagementStateFailing",
+			Status: configv1.ConditionTrue,
+			Reason: fmt.Sprintf("Unsupported management state %q for %s operator", detailedSpec.ManagementState, c.operatorName),
+		})
+	}
+
+	if equality.Semantic.DeepEqual(operatorObj, originalOperatorObj) {
+		return nil
+	}
+	if _, updateErr := c.clusterOperatorGetter.ClusterOperators().UpdateStatus(operatorObj); err != nil {
```

Comment:
by doing this, you're created a case of two writers to the clusteroperator resource.  Two writers can fight, a single writer cannot.